[Feature] CSV import and export support header

chenlinzhong opened the pull request:
# Proposed changes

Issue Number: open #8487 

## Problem Summary:

- Add two new types to stream load boker load: **csv_with_names** and **csv_with_name_sand_types**
- Add two new types to export: **csv_with_names** and **csv_with_names_and_types**

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header,rebase

chenlinzhong committed:
[Feature] CSV import and export support header

cambyzju commented on be/src/http/action/stream_load.cpp:
```diff
@@ -265,8 +266,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, StreamLoadContext* ct
+        boost::iequals(format_str, BeConsts::CSV_WITH_NAMES_AND_TYPES)) {
+        ctx->header_type = format_str;
+        //treat as CSV
+        format_str = "csv";
```
"csv" also use an const define like CSV_WITH_NAMES?

chenlinzhong replied:
ok

chenlinzhong replied:
ok

cambyzju commented on fe/fe-core/src/main/java/org/apache/doris/load/BrokerFileGroup.java:
```diff
@@ -214,7 +215,9 @@ public void parse(Database db, DataDescription dataDescription) throws DdlExcept
             if (!fileFormat.equalsIgnoreCase("parquet")
                     && !fileFormat.equalsIgnoreCase("csv")
```
"csv" define as const like csv_with_names

chenlinzhong replied:
ok

yinzhijian commented on be/src/runtime/export_sink.cpp:
```diff
@@ -96,6 +96,13 @@ Status ExportSink::send(RuntimeState* state, RowBatch* batch) {
+        RETURN_IF_ERROR(
+                _file_writer->write(reinterpret_cast<const uint8_t*>(_t_export_sink.header.c_str()),
+                                    _t_export_sink.header.size(), &written_len));
+        _t_export_sink.header = "";
```
use a flag  to indicate that the header has been sent may be better?

chenlinzhong replied:
ok

chenlinzhong committed:
[Feature] CSV import and export support header

yinzhijian commented on be/src/runtime/export_sink.cpp:
```diff
@@ -96,6 +97,13 @@ Status ExportSink::send(RuntimeState* state, RowBatch* batch) {
+        RETURN_IF_ERROR(
+                _file_writer->write(reinterpret_cast<const uint8_t*>(_t_export_sink.header.c_str()),
+                                    _t_export_sink.header.size(), &written_len));
+        _send_header = true;
```
use _header_sent may be better

chenlinzhong replied:
ok

yinzhijian commented on be/src/runtime/export_sink.h:
```diff
@@ -77,6 +77,7 @@ class ExportSink : public DataSink {
     RuntimeProfile::Counter* _bytes_written_counter;
     RuntimeProfile::Counter* _rows_written_counter;
     RuntimeProfile::Counter* _write_timer;
+    bool _send_header;
```
_header_sent may be better

chenlinzhong replied:
ok

yinzhijian commented on docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md:
```diff
@@ -78,7 +78,7 @@ under the License.
         exec_mem_limit: 导入内存限制。默认为 2GB。单位为字节。
 
-        format: 指定导入数据格式，默认是csv，支持json格式。
+        format: 指定导入数据格式，默认是csv，支持json格和带header的csv格式: csv_with_names、csv_with_names_and_types。
```
支持json格式， 少个式

chenlinzhong replied:
ok

yinzhijian commented on fe/fe-core/src/main/java/org/apache/doris/load/ExportJob.java:
```diff
@@ -252,10 +253,40 @@ private void genExecFragment() throws UserException {
+                }
+                header = header.substring(0, header.length() - getColumnSeparator().length());
+                header = header + getLineDelimiter();
+                //types 
```
use a common function to gen names may be better

yinzhijian commented on fe/fe-core/src/main/java/org/apache/doris/planner/ExportSink.java:
```diff
@@ -44,6 +45,15 @@ public ExportSink(String exportPath, String columnSeparator,
+        this.exportPath = exportPath;
+        this.columnSeparator = columnSeparator;
+        this.lineDelimiter = lineDelimiter;
+        this.brokerDesc = brokerDesc;
```
ExportSink(exportPath, columnSeparator, lineDelimiter, brokerDesc); may be better

chenlinzhong replied:
ok

chenlinzhong committed:
[Feature] CSV import and export support header

chenlinzhong committed:
[Feature] CSV import and export support header

yinzhijian reviewed (commented):
LGTM

cambyzju commented on be/src/runtime/file_result_writer.cpp:
```diff
@@ -216,12 +218,82 @@ std::string FileResultWriter::_file_format_to_name() {
+        switch (_output_expr_ctxs[i]->root()->type().type) {
+        case TYPE_BOOLEAN:
+        case TYPE_TINYINT:
+            types += "TINYINT";
```
1. can we use type_to_string in primitive_type.cpp?
2. the string for types, it is better the same as types name in create table SQL?

chenlinzhong replied:
ok, it's a good idea

chenlinzhong committed:
[Feature] CSV import and export support header (#8487)
1.Add two new types to stream load boker load: csv_with_names and csv_with_name_sand_types
2.Add two new types to export: csv_with_names and csv_with_names_and_types

cambyzju commented:
+1

morningman reviewed (dismissed):
LGTM

morningman reviewed (approved):
LGTM
